### ISSUE TITLE
Add support for importing components, data structures, etc. generated from an earlier Slint compiler invocation

### ISSUE THREAD

**Issue by tronical** [slint-ui/slint#7060]
This issue originates from #6196.

Example:

1) Crate A compiles `foo.slint`, which exports components, globals, data structures, etc. as well as business logic related to foo.slint.
2) Crate B compiles `main.slint`, which imports `foo.slint` (for example as a library) and wants to re-use components, etc. as well as the business logic from crate A, without generating `foo.slint`'s types again.

The objective is to enable this workflow for Rust and C++.

In order to implement this, we need to use the Slint library system as well as the ability to associate a Slint library with a Rust module or a C++ namespace, and in addition we need build system support so that cmake or cargo can locate the user's slint library just from the dependencies.

```cmake
add_library(MyreusableComponent ...)
// Doesn't generate the typical "app" entry but might generate more library/style reusable code
// This also will set various cmake target properties on MyreusableComponent ...)
slint_library_target_sources(MyreusableComponent NAME MyLibrary mylib.slint)

add_executable(my_app ...)
target_link_libraries(my_app MyreusableComponent)
// this needs to look at my_app's linked targets and locate those that contain Slint libraries, to be able
// to locate the header file, library name, etc.
slint_target_sources(my_app main.slint)

// main.slint can do `import { Foo } from "@MyLibrary"`
```

`my-library`'s `build.rs`:
```rust
fn main() {
    let config = slint_build::CompilerConfiguration::new().as_library("MyLibrary").rust_module("mymodule");
    slint_build::compile_with_config("mylib.slint", config);
}
```

`my-library''s `lib.rs`:
```rust
pub mod mymodule {
    slint::include_modules!()
}
```

App's `Cargo.toml`:
```toml
[dependencies]
my-library = { ... }
```

App's invocation of the slint compiler would use `DEP_MY_LIBRARY_SLINT_XXX` to locate the path to the .slint library file, name of the library, etc.

It would be tempting to use the `links` feature of Cargo for this, but that's a hack. The recommended way would be that my-library also has a `my-library-build` crate that provides API for App's `build.rs`, to then invoke the slint compiler accordingly.

**Comment by ogoffart** [comment]
#9329 added experimental support for it, but this is not finished

**Comment by karmakarmeghdip** [comment]
Based on the current example shown in /tests/manual/module-builds, it doesn't seem possible to make something that uses multiple threads, like for example, a web image component. Since the global references aren't Send. Also, it would be nicer if the user didn't need to do anything in the Rust side for using the component, i.e. more self contained and isolated.
Currently using a component, imported from a different crate, in multiple places will require manual handling. It would be nice if they were more isolated

**Comment by LeonMatthes** [comment]
Note: Cargo is experimenting with a system for DEP keys that does not require the links key 🚀 
https://github.com/rust-lang/cargo/issues/3544

Getting rid of the links key would make this a seamless experience for Rust users!

**Comment by LeonMatthes** [comment]
Discussion point: How should this deal with (bundled) translations? Should it be possible to bundle translations per-library, or only per-app?
per-library translation files might make sense, but maybe bundling/loading still always needs to happen at the app level.

**Comment by tronical** [comment]
I think that has to happen on a per-library basis, as only during the compilation we have access to the catalogues, and the translations are likely contextually bound to the library anyway. It's similar to embedded glyphs and textures: I'd say we don't worry about merging them but keep them on a per-library level.

**Comment by LeonMatthes** [comment]
> I think that has to happen on a per-library basis, as only during the compilation we have access to the catalogues, and the translations are likely contextually bound to the library anyway. It's similar to embedded glyphs and textures: I'd say we don't worry about merging them but keep them on a per-library level.

I agree with this in principle, for bundling this is also likely straightforward.
I'm unsure about runtime translations though, which are likely distributed by the app and loaded by a single call to `slint::init_translations!(...);`, so this will need a bit of hashing out.
